### PR TITLE
Add coverage build type, organize cmake

### PR DESCRIPTION
**Pull request by Warchant**
- Organize cmake in a better structure - analysis, config types.
- Add new CMAKE_BUILD_TYPE=Coverage -- it installs all necessary flags for coverage and should be used in CI.
- Remove unused fuzzing
- Organize flags
- Add ccache support for xcode, ninja, and linker

### CODE REVIEW

File: cmake/Modules/Findgrpc.cmake [hunk 1]
```diff
@@ -34,29 +34,50 @@ set(URL https://github.com/grpc/grpc)
 set(VERSION bfcbad3b86c7912968dc8e64f2121c920dad4dfb)
 set_target_description(grpc "Remote Procedure Call library" ${URL} ${VERSION})
 
+iroha_get_lib_name(GPRLIB       gpr               SHARED)
```
Comment (l4l) [inline cmake/Modules/Findgrpc.cmake]: Why is this here?

File: cmake/Modules/Findgrpc.cmake [hunk 2]
```diff
@@ -94,5 +115,6 @@ set_target_properties(gpr PROPERTIES
 if(ENABLE_LIBS_PACKAGING)
   add_install_step_for_lib(${grpc_LIBRARY})
   add_install_step_for_lib(${grpc_grpc++_LIBRARY})
+  add_install_step_for_lib(${grpc_grpc++_reflection_LIBRARY})
```
Comment (l4l) [inline cmake/Modules/Findgrpc.cmake]: Why is this here? I think we can remove it, ping @lebdron

File: cmake/Modules/Findpqxx.cmake [hunk 3]
```diff
@@ -15,27 +15,31 @@ set(URL https://github.com/jtv/libpqxx.git)
...
 
+iroha_get_lib_name(PQXX pqxx STATIC)
+
 if (NOT pqxx_FOUND)
   externalproject_add(jtv_libpqxx
       GIT_REPOSITORY ${URL}
       GIT_TAG        ${VERSION}
-      CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env CXXFLAGS=${CMAKE_CXX_FLAGS} CPPFLAGS=-I${postgres_INCLUDE_DIR} PG_CONFIG=${pg_config_EXECUTABLE} ./configure --disable-documentation --with-postgres-include=${pq_INCLUDE_DIR} --with-postgres-lib=${pq_INCLUDE_DIR}
+      CONFIGURE_COMMAND
+          ${CMAKE_COMMAND} -E env CXXFLAGS=${CMAKE_CXX_FLAGS} CPPFLAGS=-I${postgres_INCLUDE_DIR} PG_CONFIG=${pg_config_EXECUTABLE} ./configure --disable-documentation --with-postgres-include=${pq_INCLUDE_DIR} --with-postgres-lib=${pq_INCLUDE_DIR}
```
Comment (l4l) [inline cmake/Modules/Findpqxx.cmake]: Not really good idea to pass all of the flags, e.g: if I'd like to use sanitizers I don't want more performance issues

File: cmake/config/all.cmake [hunk 4]
```diff
@@ -0,0 +1,15 @@
+set(CMAKE_CXX_STANDARD 14)                    # force std=c++14
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+SET(CMAKE_POSITION_INDEPENDENT_CODE    TRUE)  # -fPIC
```
Comment (l4l) [inline cmake/config/all.cmake]: Use consistent (lower) letter-case

File: cmake/config/coverage.cmake [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#### -DCMAKE_BUILD_TYPE=Coverage
+
+# https://github.com/bilke/cmake-modules/blob/master/CodeCoverage.cmake#L90
+SET(FLAGS
```
Comment (l4l) [inline cmake/config/coverage.cmake]: Case again

File: cmake/config/coverage.cmake [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    -O1              # first level of optimizations
+    -g               # include debug info
+    --coverage
+    -fprofile-arcs
+    -ftest-coverage
```
Comment (l4l) [inline cmake/config/coverage.cmake]: `-ftest-coverage` & `--coverage` are the same

File: cmake/config/debug.cmake [hunk 7]
```diff
@@ -0,0 +1,29 @@
+SET(FLAGS
```
Comment (l4l) [inline cmake/config/debug.cmake]: Case

File: cmake/config/debug.cmake [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    -Wextra
+    -Wno-unused-parameter
+    -Wno-deprecated-declarations
```
Comment (l4l) [inline cmake/config/debug.cmake]: Not really debug-specific, moreover you add it here:
https://github.com/hyperledger/iroha/blob/c1416b8135999ce6da45c6f746a06c29f1fbd963/cmake/config/all.cmake#L13

File: cmake/dependencies.cmake [hunk 9]
```diff
@@ -6,7 +6,16 @@ set_directory_properties(PROPERTIES
     EP_PREFIX ${EP_PREFIX}
     )
 
-# Project dependencies.
+if(CMAKE_GENERATOR MATCHES Xcode)
+  set(XCODE_EXT "/Debug")
```
Comment (l4l) [inline cmake/dependencies.cmake]: Not really a good name, better use `COMPILER_BUILD_DIR`, or smth like that

File: cmake/policies.cmake [hunk 10]
```diff
@@ -0,0 +1,3 @@
+# cmake --help-policy CMP0054
```
Comment (l4l) [inline cmake/policies.cmake]: Better describe what does the policy do

File: cmake/setup.cmake [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+include(cmake/config/coverage.cmake)
+
+# all 3rd-party dependencies
+# should be last in this section
+include(cmake/dependencies.cmake)
+
+
+
+##############################
+###        CCACHE          ###
```
Comment (l4l) [inline cmake/setup.cmake]: Why isn't it moved as well?

File: launch-cxx.in [hunk 12]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (l4l) [inline launch-cxx.in]: Should these files be in the root dir?

File: docker/develop/Dockerfile [hunk 13]
```diff
@@ -25,16 +25,23 @@ RUN apt-get -y --no-install-recommends install build-essential python-software-p
...
+        wget file unzip gdb iputils-ping vim ccache \
         gcovr cppcheck doxygen graphviz graphviz-dev; \
     apt-get -y clean
 
-RUN pip3 install --upgrade pip; \
-    pip3 install flask requests
 
-# install cmake 3.5.1
+# install curl with SSL, 7.58.0
+RUN (cd /tmp; wget  https://curl.haxx.se/download/curl-7.58.0.tar.gz; \
```
Comment (l4l) [inline docker/develop/Dockerfile]: What's the problem with the previous version?

File: cmake/dependencies.cmake (hunk 9 above)
Comment (Warchant) [reply cmake/dependencies.cmake]: semantic is "directory extention when cmake generator is xcode"

File: cmake/policies.cmake (hunk 10 above)
Comment (Warchant) [reply cmake/policies.cmake]: you can always write this command and read about it

File: cmake/setup.cmake (hunk 11 above)
Comment (Warchant) [reply cmake/setup.cmake]: moved where?

File: launch-cxx.in (hunk 12 above)
Comment (Warchant) [reply launch-cxx.in]: yes, unless you can point me the better dir for these files

File: docker/develop/Dockerfile (hunk 13 above)
Comment (Warchant) [reply docker/develop/Dockerfile]: Default CURL from APT is compiled without SSL support, and Hunter (which naturally uses CURL to download binary distributions from github) fails with error.

File: cmake/Modules/Findgrpc.cmake (hunk 1 above)
Comment (Warchant) [reply cmake/Modules/Findgrpc.cmake]:
```
root@b7b6b07ca39c:/# ldd irohad\
>
	linux-vdso.so.1 =>  (0x00007ffdbc7fd000)
	libtbb.so.2 => not found
	libgrpc++.so => /usr/local/lib/libgrpc++.so (0x00007fe5af5ec000)
	libpq.so.5 => not found
	libboost_filesystem.so.1.65.1 => /usr/local/lib/libboost_filesystem.so.1.65.1 (0x00007fe5af3d1000)
	libboost_system.so.1.65.1 => /usr/local/lib/libboost_system.so.1.65.1 (0x00007fe5af1cc000)
	libprotobuf.so => /usr/local/lib/libprotobuf.so (0x00007fe5ae9da000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe5ae7bd000)
	libed25519.so.1.2.2 => /usr/local/lib/libed25519.so.1.2.2 (0x00007fe5ae5a5000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fe5ae223000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe5adf1a000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fe5add04000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe5ad93a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fe5af929000)
	libgrpc.so => not found
	libgpr.so => not found
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fe5ad720000)
```

File: cmake/Modules/Findgrpc.cmake (hunk 2 above)
Comment (Warchant) [reply cmake/Modules/Findgrpc.cmake]: Probably reflection can be removed

File: cmake/policies.cmake (hunk 10 above)
Comment (l4l) [reply cmake/policies.cmake]: When you're reading source files, you're probably don't want to switch to the terminal and read big mans

File: cmake/setup.cmake (hunk 11 above)
Comment (l4l) [reply cmake/setup.cmake]: Separate file, like analysis/*cmake

File: launch-cxx.in (hunk 12 above)
Comment (l4l) [reply launch-cxx.in]: I'm not sure that I understand their purposes

File: cmake/setup.cmake (hunk 11 above)
Comment (Warchant) [reply cmake/setup.cmake]: ccache is not an analysis tool

File: launch-cxx.in (hunk 12 above)
Comment (Warchant) [reply launch-cxx.in]: look at ccache, there is a link

File: cmake/setup.cmake (hunk 11 above)
Comment (l4l) [reply cmake/setup.cmake]: Like files in analysis, not into
I think `cmake/ccache.cmake` would be nice

File: cmake/dependencies.cmake (hunk 9 above)
Comment (l4l) [reply cmake/dependencies.cmake]: My first association is a file extension, moreover it used like that

File: cmake/setup.cmake (hunk 11 above)
Comment (Warchant) [reply cmake/setup.cmake]: Okay, I will fix that.

Comment (Warchant) [reply cmake/setup.cmake]: fixed

File: cmake/config/coverage.cmake [hunk 14]
```diff
@@ -0,0 +1,43 @@
+#### -DCMAKE_BUILD_TYPE=Coverage
+
+# https://github.com/bilke/cmake-modules/blob/master/CodeCoverage.cmake#L90
+set(FLAGS
+    -O1              # first level of optimizations
+    --coverage
+
```
Comment (l4l) [inline cmake/config/coverage.cmake]: Empty string, is it intended? Same with the following files

File: cmake/config/debug.cmake [hunk 15]
```diff
@@ -0,0 +1,22 @@
+set(FLAGS
+    -Wextra
+    -O0                             # no optimizations
```
Comment (l4l) [inline cmake/config/debug.cmake]: Can be -Og

File: cmake/release.cmake [hunk 16]
```diff
@@ -10,9 +10,9 @@ endif()
 set(CPACK_PACKAGE_NAME                "iroha")
 set(CPACK_PACKAGE_VENDOR              "Soramitsu LLC")
 set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Iroha - simple decentralized blockchain")
-SET(CPACK_PACKAGE_VENDOR              "Soramitsu LLC")
-SET(CPACK_RESOURCE_FILE_LICENSE       "${CMAKE_SOURCE_DIR}/LICENSE")
-SET(CPACK_PACKAGE_CONTACT             "Bogdan Vaneev <bogdan@soramitsu.co.jp>")
+set(CPACK_PACKAGE_VENDOR              "Soramitsu LLC")
```
Comment (l4l) [inline cmake/release.cmake]: Ltd, actually

File: cmake/config/coverage.cmake (hunk 14 above)
Comment (Warchant) [reply cmake/config/coverage.cmake]: Yes, it is a separator

**Comment by Warchant** [comment]
Build fails because of https://soramitsu.atlassian.net/browse/IR-1022. As soon as it is fixed, this PR can be merged.

File: CMakeLists.txt [hunk 17]
```diff
@@ -1,35 +1,28 @@
...
-  set(REPORT_DIR ${CMAKE_BINARY_DIR}/reports)
-  file(MAKE_DIRECTORY ${REPORT_DIR})
-  include(cmake/analysis.cmake)
-endif()
+set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
+
+#########################
+## configuration types ##
+#########################
+set(CMAKE_CONFIGURATION_TYPES "Debug;Release;Coverage" CACHE STRING "" FORCE)
```
Comment (lebdron) [inline CMakeLists.txt]: It is not required to modify CMake targets, because coverage only modifies flag set - it can be used both in Debug and Release.

File: cmake/functions.cmake [hunk 18]
```diff
@@ -148,3 +151,14 @@ function(JOIN VALUES GLUE OUTPUT)
   string (REGEX REPLACE "[\\](.)" "\\1" _TMP_STR "${_TMP_STR}") #fixes escaping
   set (${OUTPUT} "${_TMP_STR}" PARENT_SCOPE)
 endfunction()
+
+# if compiler supports flag FLAG1, then OUT=FLAG1, else OUT=FLAG2
+function(test_flag_support_or_fallback OUT FLAG1 FLAG2)
+  include(CheckCXXCompilerFlag)
+  check_cxx_compiler_flag(${FLAG1} has_${FLAG1})
```
Comment (lebdron) [inline cmake/functions.cmake]: Same check should be applied to second flag.

File: docker/develop/Dockerfile (hunk 13 above)
Comment (lebdron) [reply docker/develop/Dockerfile]: Is there a way to disable SSL in Hunter? It's unlikely that users will have curl with SSL by default.

File: docker/develop/Dockerfile [hunk 19]
```diff
@@ -25,16 +25,23 @@ RUN apt-get -y --no-install-recommends install build-essential python-software-p
...
+    tar -xf curl-*.tar.gz; \
+    cd curl-7.58.0; \
+    ./configure --with-ssl; \
+    make -j${PARALLELISM} install); \
+    ldconfig; \
+    rm -rf /tmp/curl*
+
+# install cmake 3.10.2
```
Comment (lebdron) [inline docker/develop/Dockerfile]: Having a greater version can introduce build inconsistencies, because features from versions greater than 3.7 can be used without CI failure. Having 3.7 version in Dockerfile is fine.

File: iroha-cli/CMakeLists.txt [hunk 20]
```diff
@@ -61,7 +61,6 @@ target_link_libraries(iroha-cli
     client
     cli-flags_validators
     keys_manager
-    boost
```
Comment (lebdron) [inline iroha-cli/CMakeLists.txt]: How are these boost changes related to pull request description?

File: CMakeLists.txt (hunk 17 above)
Comment (Warchant) [reply CMakeLists.txt]: It is convenient to group flags as configuration types. The main reason is that you can select specific configuration at cmake time 

```
cmake .. -DCMAKE_BUILD_TYPE=Coverage
```

File: cmake/functions.cmake (hunk 18 above)
Comment (Warchant) [reply cmake/functions.cmake]: Why? It is a shortcut for 
`OUT=defined(flag1) ? flag1 : flag2`

By definition, flag2 should be supported

File: iroha-cli/CMakeLists.txt (hunk 20 above)
Comment (Warchant) [reply iroha-cli/CMakeLists.txt]: I will remove them. Actually all these boost targets are not needed, everything works without them.

File: docker/develop/Dockerfile (hunk 19 above)
Comment (Warchant) [reply docker/develop/Dockerfile]: @lebdron what is the exact cmake version required for SWIG?

File: docker/develop/Dockerfile (hunk 13 above)
Comment (Warchant) [reply docker/develop/Dockerfile]: It is possible to disable SSL https://docs.hunter.sh/en/latest/faq/how-to-fix-download-error.html?highlight=SSL but actually, SSL issue is mostly related to docker. If you have a desktop linux version or mac, it is likely to have libssl or libboringssl and libcurl with SSL support.

**Comment by Warchant** [comment]
This PR will be split onto multiple parts.
